### PR TITLE
Make metrics logging work for pipeline parallelism

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #383

Avoid complicating the ux and leave the status quo of 2 user-selectable
behaviors:
 - log from rank 0 (the default)
 - log from all ranks (not the default)

Modify the meaning of 'log from rank 0' to log from rank 0 in
non-pipeline parallel runs, and log from the local rank 0 within the
last pipeline-parallel stage group if pp is enabled.  (note: earlier
pipeline stages still produce some metrics like mfu/memory, but do not
compute loss.)

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -244,7 +244,12 @@ def loss_fn(pred, labels):
     optimizer = build_optimizer(model, job_config)
     scheduler = get_lr_scheduler(optimizer, job_config)
 
-    metric_logger = build_metric_logger(job_config)
+    if parallel_dims.pp_enabled:
+        pp_size = pp_mesh.size()
+        metrics_log_rank = int((world_mesh.size() // pp_size) * (pp_size - 1))
```
Comment (awgu) [inline train.py]: Does this rank computation assume that PP is outermost? If so, should we assert/check for that?

Comment (wconstab) [reply train.py]: yea, it does. I wonder if there is a better place to do the assert.  I'll add it here for now.

I also don't like to do it this way, i might want to propose adding a device-mesh API to help do a calculation like this more robustly.

Comment (wconstab) [reply train.py]: i moved this into a util and added an assert.

**Review by fegin — APPROVED** [review]
We probably need to extend DeviceMesh to make calculating a specific rank easier.

**Comment by wconstab** [comment]
> We probably need to extend DeviceMesh to make calculating a specific rank easier.

yes, my thoughts exactly. I would like to discuss this offline. I couldn't quickly think of what the best API proposal for devicemesh would be so i went this route instead.
